### DISCUSSION TITLE
<**> and lift2

### DISCUSSION THREAD

**Discussion by mihneagogu** [j-mie6/Parsley#41] [Q&A] [answered]
Hey, how would one use `<**>` ?
I want the result of two integer parsers to multiply, which is very easy with plain `lift2`:
`lift2(_ * _, intParser1, intParser2)`. How would one write this using `<**>` ?
I found this:
```scala
val intMul = (x: Int) => (y: Int) => x * y
val mulParser = intParser1 <**> (intParser2.map(intMul) </> identity[Int])
```
Is there a better way to do this?
Edit: `val mulParser = intParser1 <**> (intParser2 <#> intMul)` seems to do the trick. Is this the correct way to use `<**>` ?

**Comment by j-mie6** [comment, marked as answer]
Yeah, `val mulParser = intParser1 <**> intParser2.map(intMul)` will do the trick, but you probably want to define:

```scala
val intMul = (y: Int) => (x: Int)  => x * y
```

So that the arguments are reversed, that way it correctly models what you had with the `lift2`: obviously, for an associative operation like `*` it doesn't actually matter.

**Reply by j-mie6** [reply]
By the way, the snippet that you posted is used if you had something like this:

```scala
val mulParser = attempt(lift2(_*_, intParser1, intParser2)) <|> intParser1
```

And you wanted to factor out the leading `intParser1` to remove any backtracking.
